### ISSUE TITLE
[RFE] Certificate fingerprint verification

### ISSUE THREAD

**Issue by jirib** [OpenSMTPD/OpenSMTPD#444]
All TLS trust is problematic (distribution of CA, revokation), verification of specific certificate finterprint could solve the issue as it doesn't need to trust whole chain... See http://postfix.1071664.n5.nabble.com/TLS-Certificate-signature-failure-what-is-the-reason-tp62435p62484.html

**Comment by poolpOrg** [comment]
work has began on switching OpenSMTPD to libtls which will make this feature request trivial to implement.

stay tuned

**Comment by poolpOrg** [comment]
So, this has become trivial with the libtls conversion:

    8afeb1d4fe8f7bb9 mta tls ciphers=TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256
    8afeb1d4fe8f7bb9 mta cert-check result="failure" fingerprint="SHA256:10b7bd2bc6bc1ec0eda97bef7fad8fc78e8bb6d9a9d6bb48faa63a7ec26bc9ad"

All we need to do is add an optional fingerprint table to relay rules and perform a lookup table with the fingerprint as the key if that table is present, bypassing CA check in this case, that's probably less than 50 lines of code in total in my libtls branch.

BUT...

The libtls conversion will not be merged to master before a couple release at least and we have the MTA layer rewrite project which will void any work I do in that area and force to me to do it again.

I'll put this ticket on hold until we have the new MTA.

**Comment by ngortheone** [comment]
so libtls work was done in https://poolp.org/posts/2020-01-22/january-2020-opensmtpd-work-libasr-and-libtls/

Does it unblock this issue? Or are we still waiting for MTA rewrite?

**Comment by poolpOrg** [comment]
It makes it easier to handle in the MTA but it still depends on the MTA rewrite.
